### PR TITLE
Added module for making SOSS simulations with random contaminants

### PR DESCRIPTION
**Pull request by hover2pi**
Here's a module that will make the simulations and write the results out to an .h5 file.

### CODE REVIEW

**Comment by hover2pi** [comment]
@hstewart93 I just added a function `find_order0s` that lets you point to a SOSS 'rate.fits' file and it will spit out a table of the sources in the field along with an interactive plot where you can hover over the contaminants and it shows you the Gaia EDR3 source info. We can use this to find order 0 traces of different Teff values to assemble a trace library.

**Comment by hover2pi** [comment]
DISCLAIMER: I didn't test the `find_order0s` function yet, haha. I'll try it on some data in a little bit, e.g. WASP-93 and WASP-17.

File: simulations/soss_simulations.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+    Parameters
+    ----------
+    rate_file: str
+        The rate file of the observation to load
+    plot: bool
+        Plot the data along with the identified order 0 contaminants
+    """
+    # Get the data
+    data = fits.getdata(rate_file)
```
Comment (hstewart93) [inline simulations/soss_simulations.py:43]: Is this astropy fits, it hasn't been imported so wanted to check before adding the import statement

**Comment by hstewart93** [comment]
Thanks so much for this @hover2pi! I have added another function `run_simulations_random` which is a copy of your function but with the Teff and Jmag of the target being randomly generated. Currently the code randomly generates contaminants but their positions appear to be fixed:

image

Is it possible to randomise the positions please? I had a look but couldn't figure out what needed changing. Thanks!

File: simulations/soss_simulations.py (hunk 1 above)
Comment (hover2pi) [reply simulations/soss_simulations.py:43]: Good catch!

**Comment by hover2pi** [comment]
Oh, yeah, I didn't test very well, haha. Yeah, I can fix this...

**Comment by hover2pi** [comment]
Ok, I updated the original function to vary the target parameters and now it prints out a table as a sanity check. E.g.
`ss.run_simulations(targ_Jmag=(0, 10), targ_Teff=(2000, 10000), N_contaminants=(0, 10))`

**Comment by hover2pi** [comment]
Screenshot 2024-12-16 at 12 07 33 PM

**Comment by hover2pi** [comment]
I just realized that the pickoff mirror starts at col 700 and there are no order 0s before that. I updated the code to only place order 0s in cols 700-2048.

**Comment by hstewart93** [comment]
Thanks for making those updates @hover2pi, the sims are looking great! Is it possible to randomise the positions of the contaminants for each spectra generated?

**Comment by hover2pi** [comment]
When you run `ss.run_simulations(targ_Jmag=(0, 10), targ_Teff=(2000, 10000), N_contaminants=(0, 10))` they aren't randomized?

**Comment by hstewart93** [comment]
These are the spectra I get when running `ss.run_simulations(targ_Jmag=(0, 10), targ_Teff=(2000, 10000), N_contaminants=(0, 10))`:

image

The number of contaminants are random, but the positions are the same.

**Comment by hover2pi** [comment]
Maybe the file isn't being overwritten? I just tried it with your plotting calls in `simulated_data.ipynb` and it looks right to me. Let me know if something is amiss!
download

**Comment by hstewart93** [comment]
Hmm I wonder if this is a versioning issue, could you tell me what Python version you are using and the versions of any packages please so I can check against mine?

**Comment by hover2pi** [comment]
Since it uses ExoCTK, I am just using the ExoCTK Python 3.11 env here: https://github.com/ExoCTK/exoctk/blob/main/env/environment-3.11.yml

**Comment by hstewart93** [comment]
@hover2pi unrelated but a weird intermittent error `OSError: Header missing END card.` when running for many (>100) simulations. I have found the route of the issue and it can be resolved by setting `targ_Teff=(2000, 7000)`. It seems to be caused by a corrupt fits file in the `exoctk_contam` directory. When the Teff is above 7000, exoctk attempts to open the file `NIS_SOSS_7100.fits` as the nearest effective temp. Below is the output of fitsheader for that file. I have downloaded the exoctk data directly from the link https://stsci.box.com/s/7ph64s6cfyusfcxjvih8ll5rn0ydzw86 and redownloaded to double check nothing happened during download. For now I can simply set `targ_Teff=(2000, 7000)` as a workaround to generate data for developing the neural net. I will open an issue in exoctk about the fits file. Thanks!

```
fitsheader /data/typhon2/hattie/exoctk_data/exoctk_contam/traces/NIS_SUBSTRIP256/NIS_SOSS_7100.fits
WARNING: non-ASCII characters are present in the FITS file header and have been replaced by "?" characters [astropy.io.fits.util]
WARNING: Header block contains null bytes instead of spaces for padding, and is not FITS-compliant. Nulls may be replaced with spaces upon writing. [astropy.io.fits.header]
ERROR: Header missing END card. [astropy.io.fits.scripts.fitsheader]
```

**Comment by hstewart93** [comment]
Also I found the cause of the contaminants being placed not randomly. The numpy random number generator is not thread safe, so each process was repeating the same numbers as the others. Setting `max_workers` to 1 resolves this as the state is randomised for 1 core, but is duplicated for multiple threads. Switching to built in Python `random` (which is thread safe) fixed it. I think it's a play off between number of processes vs number of cores as to whether the lack of randomness would be apparent, as I was running only 10 processes on 128 cores there was no randomness in the positions. Very elusive!

**Comment by hstewart93** [comment]
@hover2pi PR for the changes here: https://github.com/hstewart93/jwst/pull/2 if you don't mind giving it a quick look please

**Pull request merged**